Implement Tensor.__cuda_array_interface__.

asford opened the pull request:
_Implements pytorch/pytorch#11914, cc: @ezyang_

Implements `__cuda_array_interface__` for non-sparse cuda tensors,
providing compatibility with numba (and other cuda projects...).

Adds `numba` installation to the `xenial-cuda9` jenkins test environments via direct installation in `.jenkins/pytorch/test.sh` and numba-oriented test suite in `test/test_numba_integration.py`.

See interface reference at:
https://numba.pydata.org/numba-doc/latest/cuda/cuda_array_interface.html

Commit:
Add Tensor.__cuda_array_interface__ test.

Add baseline test of python-level __cuda_array_interface__ for
torch.Tensor.

See interface reference at:
https://numba.pydata.org/numba-doc/latest/cuda/cuda_array_interface.html

Commit:
Implement Tensor.__cuda_array_interface__

Commit:
Delint __cuda_array_interface__ implementation.

ezyang commented on torch/tensor.py:
```diff
@@ -441,4 +441,52 @@ def __array_wrap__(self, array):
+                "Use var.detach().__cuda_array_interface__ instead."
+            )
+
+        typestr = {
```
Aren't you obligated to specify endianness in the type string?

asford replied:
This is currently working because numba assumes the native byte order of the device if it isn't specified, though I agree that this (may) a bug in my implementation.

I'm not super familiar with the `pytorch` internals and don't have access to a big-endian system, so please feel free to correct me, but I believe:

* `pytorch` currently does not support non-native byte orders when [converting from numpy](https://github.com/asford/pytorch/blob/e7c7a45c620cd3d8f76e3f3906e557d7737d0c12/torch/csrc/utils/tensor_numpy.cpp#L124-L128).
* `pytorch` does not have a way of programmatically accessing the byteorder of tensor structures, which are all represented in native ordering.
* Numba converts dtype representations via [`numpy.dtype`](https://github.com/numba/numba/blob/master/numba/cuda/api.py#L194-L204), which will use the host-endianness if explicitly native ("=") or not specified (as in this implementation).
* Cuda devices are little-endian, as are `x86` and `ppc64le` deployments.

I think the options are:

1. Specify all little-endian via "<".
2. Specify 'native' byte ordering via "=", which will currently break if there was some form of `ppc64`/`cuda` cross deployment but _might_ more accurately reflect the semantics of the underlying buffer.

I will update this to return "<", unless there's a missed assumption above.

ezyang replied:
I'm definitely not saying that anyone is actually using big-endian, but it's simply a matter of hewing to the standard. "<" seems fine.

ezyang commented on test/test_torch.py:
```diff
@@ -8426,6 +8426,89 @@ def test_numpy_index(self):
             self.assertFalse(isinstance(idx, int))
             self.assertEqual(x[idx], x[int(idx)])
 
+    @unittest.skipIf(not TEST_NUMPY, "Numpy not found")
```
So... I suppose this test is a reasonable smoketest, but what I *really* want tested, is that now a Torch tensor can successfully be used by the numba library. I don't see that tested here, and that doesn't seem good!

ezyang replied:
You should test if numba is available before running said test, because not all test environments will have numba available (CI for one).

asford replied:
👍 Sounds great to me, I'd also prefer to have this tested on the pytorch-side.

I will port a few internal tests that test this integration and handle the numba library detection in a manner similar to how scipy is currently detected.

asford replied:
I've pushed a set of baseline numba integration cases which are "working on my machine" 😉.

This will likely require adding numba to your CI configuration, presumably at pietern/pytorch-dockerfiles? Would you like to coordinate a change to your CI before this pull?

ezyang replied:
You can test out adding it yourself by conda install'ing numba in `.jenkins/pytorch/test.sh`. Do it in only one configuration, e.g., one of our conda CUDA ones.

ezyang reviewed (changes requested):
Lack of actual numba test is blocker in my mind. I have a few other doc nits which I'll mention below.

ezyang commented on torch/tensor.py:
```diff
@@ -441,4 +441,52 @@ def __array_wrap__(self, array):
+        # hasattr(cpu_tensor, "__cuda_array_interface__") is False.
+        if not self.device.type == "cuda":
+            raise AttributeError(
+                "Not on cuda device, use Tensor.cuda() first: %r" %
```
The suggestion to use cuda() seems very odd from my perspective, because I shouldn't be calling __cuda_array_interface__ unless I *knew* that the tensor was a CUDA tensor. If I am calling it on a non-CUDA tensor... well, there is certainly an error, but the fact that you don't have a CUDA tensor doesn't seem like the most salient one.

asford replied:
🤔 I chose this phrasing to roughly match the wording of `Tensor.__array__()`, which raises with the message:

`RuntimeError: can't convert CUDA tensor to numpy. Use tensor.cpu() to copy tensor to host memory first.`

I've found that a cpu-tensor-provided-to-cuda-operation (or the inverse) is one of the more common bug types when working with torch, as folks (a) may be coming from libraries which implicitly copy cross device and/or (b) may not fully implement the "allocate on device of your input tensor" design pattern. 

The current behavior of numba is to perform implicit cross-device copies when provided a numpy array and the difference in behavior when moving from `ndarray` to `tensor` may be the first pain point a end-user experiences. For these folks having a explicit callout to `tensor.cuda()` to perform the cross-device copy seems useful. 

More sophisticated users will likely have a deeper understanding of the memory-management semantics of the library, and understand that wantonly running cross-devices transfers isn't a great idea. I think exact exception message will matter less; they'll recognize the cross-device mismatch and begin determining how the issue arose upstream of the call point.

Would you be open to just clarifying the wording to more explicitly matching the `__array__` message?

`AttributeError: can't convert {device.type} tensor to cuda array view. Use tensor.cuda() to copy tensor to device memory first.`

I'm open to alternatives, but I'm not sure what a better phrasing for the exception message would be.

ezyang replied:
That message sounds better.

ezyang commented on torch/tensor.py:
```diff
@@ -441,4 +441,52 @@ def __array_wrap__(self, array):
+        if self.requires_grad:
+            raise RuntimeError(
+                "Can't get __cuda_array_interface__ on Variable that requires grad. "
+                "Use var.detach().__cuda_array_interface__ instead."
```
If our user blindly follows this advice without knowing what detach() actually does, they might come calling us when their code didn't work, because this advice didn't solve the fundamental problem which is that you can't autograd past PyTorch. More robust advice would be to say, "If gradients are truly not required, you can get a Variable which doesn't require grad by saying var.detach()". And I wouldn't advise the user to use __cuda_array_interface__, because in all likelihood this call is happening in code out of their control.

asford replied:
:+1: As before, this was an attempt to match `Tensor.numpy()`, but I agree with your points.

How about:

```
"Can't get __cuda_array_interface__ on Variable that requires grad. "
"If gradients are not required, use var.detach() to get Variable that does not require grad."
```

ezyang replied:
I would probably also say to update `Tensor.numpy()` too :) SGTM

asford replied:
To follow up, this standard message is used across a relatively wide collection of tensor interfaces when guarding against non-gradient-safe calls. I'm happy to do the update across the board if you'd like, but want to double check before I touch a wider chunk of the interface.

Ex: [numpy](https://github.com/asford/pytorch/blob/36019decacf872dfa48d99d990cf1a0cb9b257ef/tools/autograd/templates/python_variable_methods.cpp#L327-L328), [map_](
https://github.com/asford/pytorch/blob/36019decacf872dfa48d99d990cf1a0cb9b257ef/tools/autograd/templates/python_variable_methods.cpp#L400-L401), [map2_](https://github.com/asford/pytorch/blob/36019decacf872dfa48d99d990cf1a0cb9b257ef/tools/autograd/templates/python_variable_methods.cpp#L418-L419) et al...

Commit:
Update cuda array interface exception messages.

Commit:
Add byte order to cuda array typestr.

Commit:
Add initial numba integration tests.

Add initial pytorch/numba integration tests. Expand test
common/common_cuda configuration to support optional numba installation.
Add test/test_numba_integration.py, with baseline integration test
cases.

Commit:
Tidy numba tests, add conversion error cases.

Commit:
Fix test error in test_autograd.

Commit:
Add numba to xenial-cuda9 tests.

Install numba on xenial-cuda9 test runs, allowing integration tests.

Commit:
Enable numba_integration test suite.

Commit:
Fix py2/3 behavior delta in numba integration test.

Fix numba __cuda_array_interface__ integration check to resolve
difference in py2/py3 hasattr behavior. Tidy documentation.

Commit:
Delint.

apaszke commented on test/test_autograd.py:
```diff
@@ -2165,6 +2165,10 @@ def test_dir(self):
 
+        # __cuda_array_interface__ attribute is not valid for non-cuda tensors,
+        # Marks this via "AttributeError", hasattr is False
+        keys.remove("__cuda_array_interface__")
```
Can we fix `dir` to not return it for CPU tensors then? This test is there for a reason

Commit:
Fix __dir__ implementation for __cuda_array_interface__.

Update `torch.Tensor.__dir__` to only include `__cuda_array_interface__`
if property will be present (i.e. a dense, cuda tensor). Preserved
RuntimeError behavior of cuda tensors that require grad.

Commit:
Pin numba version in test environment.

asford commented:
@ezyang

This is ready for another round of review when you get the chance. I've:

1. Expanded the test suite to cover baseline numba integration and added numba installation to xenial-cuda9 py2 & py3 test configurations.

    This _could_ be expanded into a full end-to-end test, covering dispatch of a jit-compiled numba function over input torch.Tensors and/or a full test of autograd integration via numba-jitted forward/backward passes. I'm not entirely sure about the project's philosopy of what should live in the test layer vs examples, and I'd be happy to hear your thoughts.
    
2. Tidied up the __cuda_array_interface__ `dir` behavior.
2. Tidied up the __cuda_array_interface__ error messages.

I'm a _bit_ concerned about adding the numba installation to your test runner, rather than folding the numba package into your test image. I've pinned to installed version to attempt to mitigate the risk of a dependency on an external package, but this still adds ~400mb of external package access to the test run.

ezyang commented:
> I'm a bit concerned about adding the numba installation to your test runner, rather than folding the numba package into your test image. I've pinned to installed version to attempt to mitigate the risk of a dependency on an external package, but this still adds ~400mb of external package access to the test run.

Well, here's another crazy idea. We don't *necessarily* need to test against numba; we just need to test against something that supports cuda_array_interface. There's no reason we couldn't make PyTorch support cuda_array_interface. So what we could do is (1) add read-in support for cuda_array_interface to PyTorch, and then (2) test that going to a dummy object with cuda_array_interface and then back again works. Then as long as we don't have correlated errors between the decoder and encoder, you have a reasonably robust test.

ezyang commented:
Other possibility: we have the numba test but don't run it in CI (only run it on a nightly build, or something.)

avmgithub commented:
@asford Would you mind if I got a version of your changes to test on ppc64le before it gets merged ?  Specially if its going to be a default test in run_test.py.   I see you require numba=0.39 cudatoolkit=9.0 for x86 but I do not know what that translates for ppc64le .   Appreciate it .   CC: @hartb @sdmonov

avmgithub commented:
@asford I did a quick test on this on my ppc64le system that does not have numba expecting all test to be skipped. But it failed in test_cuda_array_interface.   Looks like you will need to place a @unittest.skipIf(not TEST_NUMBA_CUDA, "No numba.cuda")  for it also.  The failure is :  
Traceback (most recent call last):
  File "test_numba_integration.py", line 85, in test_cuda_array_interface
    self.assertTrue(hasattr(cudat, "__cuda_array_interface__"))
AssertionError: False is not true

asford commented:
@avmgithub Great! Thinks for following up on that test. I was racking my brain trying to understand how it could be failing in that way.

asford commented:
@ezyang 

I'm comfortable with either solution you've proposed, mainly trying to avoid having a drive-by contribution that makes your testing system more fragile. I've been burned a few times by having dependencies on large external downloads in CI and think it's ideal to avoid it if possible. 

Having this run as a nightly may be best if you already have the infrastructure in place for that. Are the pytorch-examples run on a periodic basis as part of your CI?

Having the ability to read data from numba seems generally useful. You can approximate this now by creating a numbabased view of the "to" tensor and then copy data via numba from the "from" data buffer, but executing this in torch is more natural. Given that it both improves the testing situation and improves the interface I'm in favor of this approach.

Adding read support for `__cuda_array_interface__` components makes this feature _very_ similar to the existing dlpack utilities. I'm going to quickly review that implementation to make sure I'm not adding a bunch of duplicate logic.

ezyang committed:
Don't install numba

ezyang committed:
clear one more newline

ezyang commented:
> Having this run as a nightly may be best if you already have the infrastructure in place for that. Are the pytorch-examples run on a periodic basis as part of your CI?

Yes. @pjh5, do you think we could install numba when we do nightly build and tests?

> Adding read support for __cuda_array_interface__ components makes this feature very similar to the existing dlpack utilities. I'm going to quickly review that implementation to make sure I'm not adding a bunch of duplicate logic.

SGTM. In the meantime, I'm going to merge this without the CI change. Existence of test LGTM.

asford commented:
+1 @ezyang Thanks for driving this. Should we open a follow-up issue for the nightly?

ezyang commented:
Yes please.

On October 12, 2018 10:59:21 AM PDT, Alex Ford <notifications@github.com> wrote:
>+1 @ezyang Thanks for driving this. Should we open a follow-up issue
>for the nightly?
>
>-- 
>You are receiving this because you were mentioned.
>Reply to this email directly or view it on GitHub:
>https://github.com/pytorch/pytorch/pull/11984#issuecomment-429409416

-- 
Sent from my Android device with K-9 Mail. Please excuse my brevity.
